core/drlibc/drlibc_module_elf.c

```diff
@@ -480,26 +480,40 @@ elf_loader_map_phdrs(elf_loader_t *elf, bool fixed, map_fn_t map_func,
                 do_mmap = false;
                 elf->image_size = last_end - lib_base;
             }
-            /* XXX:
-             * This function can be called after dynamo_heap_initialized,
-             * and we will use map_file instead of os_map_file.
-             * However, map_file does not allow mmap with overlapped memory,
-             * so we have to unmap the old memory first.
-             * This might be a problem, e.g.
-             * one thread unmaps the memory and before mapping the actual file,
-             * another thread requests memory via mmap takes the memory here,
-             * a racy condition.
-             */
             if (seg_size > 0) { /* i#1872: handle empty segments */
                 if (do_mmap) {
-                    (*unmap_func)(seg_base, seg_size);
-                    map = (*map_func)(
-                        elf->fd, &seg_size, pg_offs, seg_base /* base */,
-                        seg_prot | MEMPROT_WRITE /* prot */,
-                        MAP_FILE_COPY_ON_WRITE /*writes should not change file*/ |
-                            MAP_FILE_IMAGE |
-                            /* we don't need MAP_FILE_REACHABLE b/c we're fixed */
-                            MAP_FILE_FIXED);
+                    if (remap_func != NULL) {
+                        /* The relevant part of the anonymous map obtained above is
+                         * expected to automatically and atomically get unmapped because
+                         * we use remap_func (which requires MAP_FILE_FIXED).
+                         */
+                        map = (*remap_func)(
+                            elf->fd, &seg_size, pg_offs, seg_base /* base */,
+                            seg_prot | MEMPROT_WRITE /* prot */,
+                            MAP_FILE_COPY_ON_WRITE /*writes should not change file*/ |
+                                MAP_FILE_IMAGE |
+                                /* we don't need MAP_FILE_REACHABLE b/c we're fixed */
+                                MAP_FILE_FIXED);
+                    } else {
+                        /* i#7192:
+                         * This function can be called after dynamo_heap_initialized,
+                         * and we will use map_file instead of os_map_file.
+                         * However, map_file does not allow mmap with overlapped memory,
```

Comment:
Where does d_r_map_file "not allow mmap with overlapped memory"?  It seems happy to take FIXED.  Is it only the heap accounting, and it doesn't "not allow" it in the sense of detecting and rejecting: it would just have the stats get off, right?  I guess all the heap routines require strict alloc-dealloc pairs.  Clarify the comment?

Reply:
> I guess all the heap routines require strict alloc-dealloc pairs.

I don't think we require strict alloc-dealloc pairs. Even in the existing code, the unmap isn't done for the exact same range allocated initially. Just that everything in the initially allocated range should have a later corresponding unmap.

Added to comment.